pkg/apiserver/registry/networkpolicy/ipassociation/rest.go

```diff
@@ -0,0 +1,137 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/apis/controlplane"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	ga "antrea.io/antrea/pkg/apiserver/registry/networkpolicy/groupassociation"
+	"antrea.io/antrea/pkg/controller/types"
+)
+
+type REST struct {
+	podQuerier   podAssociationQuerier
+	eeQuerier    externalEntityAssociationQuerier
+	ipbQuerirer  ipBlockGroupAssociationQuerier
+	groupQuerier ga.GroupAssociationQuerier
+}
+
+var (
+	_ rest.Storage = &REST{}
+	_ rest.Scoper  = &REST{}
+	_ rest.Getter  = &REST{}
+)
+
+// NewREST returns a REST object that will work against API services.
+func NewREST(podQuerier podAssociationQuerier,
+	eeQuerier externalEntityAssociationQuerier, ipbQuerier ipBlockGroupAssociationQuerier,
+	groupQuerier ga.GroupAssociationQuerier) *REST {
+	return &REST{podQuerier, eeQuerier, ipbQuerier, groupQuerier}
+}
+
+type podAssociationQuerier interface {
+	GetAssociatedPod(ip string) (*v1.Pod, error)
+}
+
+type externalEntityAssociationQuerier interface {
+	GetAssociatedExternalEntities(ip string) (*v1alpha2.ExternalEntityList, error)
+}
+
+type ipBlockGroupAssociationQuerier interface {
+	GetAssociatedIPBlockGroups(ip net.IP) ([]types.Group, error)
+}
+
+func (r *REST) New() runtime.Object {
+	return &controlplane.IPAssociation{}
+}
+
+func (r *REST) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	parsedIP := net.ParseIP(name)
+	if parsedIP == nil {
+		return nil, errors.NewBadRequest("not a valid IP address")
+	}
+	var matchedGroups []controlplane.GroupReference
+	addToGroupList := func(groups []types.Group) {
+		for _, g := range groups {
+			item := controlplane.GroupReference{
+				Namespace: g.SourceReference.Namespace,
+				Name:      g.SourceReference.Name,
+				UID:       g.UID,
+			}
+			matchedGroups = append(matchedGroups, item)
+		}
+	}
+	addAssociatedGroups := func(obj metav1.Object) {
+		groups, err := r.groupQuerier.GetAssociatedGroups(obj.GetName(), obj.GetNamespace())
+		if err != nil {
+			klog.ErrorS(err, "failed to get associated groups", "namespace", obj.GetNamespace(), "name", obj.GetName())
+		}
+		addToGroupList(groups)
+	}
+	if r.podQuerier == nil {
+		return nil, errors.NewInternalError(fmt.Errorf("traceflow must be enabled for IP association queries"))
+	}
+	matchedPod, err := r.podQuerier.GetAssociatedPod(parsedIP.String())
+	if err != nil {
+		return nil, errors.NewInternalError(err)
+	}
+	if matchedPod != nil {
+		addAssociatedGroups(matchedPod)
+	} else {
+		// The namespace/name might correspond to an ExternalEntity rather than a Pod
+		if matchedExternalEntities, err := r.eeQuerier.GetAssociatedExternalEntities(parsedIP.String()); err != nil {
+			return nil, errors.NewInternalError(err)
+		} else if matchedExternalEntities != nil && len(matchedExternalEntities.Items) > 0 {
+			for i := range matchedExternalEntities.Items {
```

Comment:
using `ExternalEntityList` as the return type is unnecessary and makes it code complexer on both caller and callee side. There is no problem to just use slice.

Reply:
Done